File: app/src/main/java/com/yvrun/officeprocess/mvp/net/DialogObserver.java
package com.yvrun.officeprocess.mvp.net;

import android.util.Log;

import com.yvrun.officeprocess.mvp.view.IBaseView;

import io.reactivex.Observer;
import io.reactivex.disposables.Disposable;

public abstract class DialogObserver<T> implements Observer<T> {

    private IBaseView mBaseView;

    public DialogObserver(IBaseView baseView) {
        mBaseView = baseView;
    }

    @Override
    public void onSubscribe(Disposable d) {

        Log.d("--TAG--","onSubscribe");
        if (mBaseView != null){
            mBaseView.showBefore();
        }
    }

    @Override
    public void onNext(T t) {

        Log.d("--TAG--","onNext");

        success(t);
    }

    @Override
    public void onError(Throwable e) {

        Log.d("--TAG--","onError");
        if (mBaseView != null){
            mBaseView.hideAfter();
        }
        error(e);
    }

    @Override
    public void onComplete() {

        Log.d("--TAG--","onComplete");
        if (mBaseView != null){
            mBaseView.hideAfter();
        }
    }

    protected abstract void success(T t);
    protected abstract void error(Throwable e);

}

File: kotlinappother/src/main/java/com/chengzhen/kotlinappother/mvp/net/DialogObserver.kt
package com.chengzhen.kotlinappother.mvp.net

import android.util.Log


import com.chengzhen.kotlinappother.mvp.view.IBaseView

import io.reactivex.Observer
import io.reactivex.disposables.Disposable

abstract class DialogObserver<T>(private val mBaseView: IBaseView?) : Observer<T> {

    override fun onSubscribe(d: Disposable) {

        Log.d("--TAG--", "onSubscribe")
        mBaseView?.showBefore()
    }

    override fun onNext(t: T) {

        Log.d("--TAG--", "onNext")

        success(t)
    }

    override fun onError(e: Throwable) {

        Log.d("--TAG--", "onError")
        mBaseView?.hideAfter()
        error(e)
    }

    override fun onComplete() {

        Log.d("--TAG--", "onComplete")
        mBaseView?.hideAfter()
    }

    protected abstract fun success(t: T)
    protected abstract fun error(e: Throwable)

}
